### PR TITLE
[AEA-1176] Add cert_request to connection.yaml

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Add cert_requests field to connectino.yaml for proof of representation

## Fixes

n/a

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comment

n/a

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -289,6 +292,157 @@ def dependencies_to_json(dependencies: Dependencies) -> Dict[str, Dict]:
...
+
+        It is expected to follow ISO 8601.
+
+        :param obj: the input to parse.
+        :return: a datetime.datetime instance.
+        """
+        result = (
+            obj
+            if isinstance(obj, datetime.datetime)
+            else datetime.datetime.fromisoformat(obj)
```
Comment (marcofavorito) [inline aea/configurations/base.py:342]: this doesn't work on Python 3.6, fixing it

File: tests/data/dummy_connection/connection.yaml [hunk 2]
```diff
@@ -27,3 +27,14 @@ dependencies:
   dep4:
     version: <=1.11.12b2
 is_abstract: false
+cert_requests:
+- identifier: cert_id
+  not_after: '2020-01-02T00:00:00'
+  not_before: '2020-01-01T00:00:00'
+  path: some/path_1
```
Comment (DavidMinarsch) [inline tests/data/dummy_connection/connection.yaml:34]: Maybe `save_path` ? To. make it obvious

File: tests/data/dummy_connection/connection.yaml [hunk 3]
```diff
@@ -27,3 +27,14 @@ dependencies:
...
+  public_key: key_id
```
Comment (DavidMinarsch) [inline tests/data/dummy_connection/connection.yaml:35]: What's `key_id` here?

File: aea/configurations/base.py [hunk 4]
```diff
@@ -289,6 +292,157 @@ def dependencies_to_json(dependencies: Dependencies) -> Dict[str, Dict]:
     return result
 
 
+class CertRequest:
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:295]: Needs to be moved to helpers class. Please review https://github.com/fetchai/agents-aea/pull/2050 before continuing (in particular, `AgentRecord`)

File: aea/configurations/base.py [hunk 5]
```diff
@@ -289,6 +292,157 @@ def dependencies_to_json(dependencies: Dependencies) -> Dict[str, Dict]:
...
+            # this raises ValueError if the input is not a valid hexadecimal string.
+            int(public_key_str, 16)
+            self._public_key = public_key_str
+            return
+
+        enforce(
+            False,
+            f"Public key field '{public_key_str}' is neither a valid identifier nor an address.",
+            exception_class=ValueError,
+        )
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:381]: we need to use crypto there

File: tests/data/dummy_connection/connection.yaml (hunk 3 above)
Comment (marcofavorito) [reply tests/data/dummy_connection/connection.yaml:36]: the crypto identifier

File: aea/cli/issue_certificates.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+@click.pass_context
+@check_aea_project
+def issue_certificates(click_context):
+    """Issue certificates for connections that require them."""
+    ctx = cast(Context, click_context.obj)
+
+    for connection_id in ctx.agent_config.connections:
+        _process_connection(ctx, connection_id)
+
+    click.echo("Done!")
```
Comment (DavidMinarsch) [inline aea/cli/issue_certificates.py:48]: A more informative error message would be good :)

File: aea/cli/issue_certificates.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+def _process_connection(ctx: Context, connection_id: PublicId):
+    click.echo(f"Processing connection '{connection_id}'...")
+    path = get_package_path_unified(ctx, CONNECTION, connection_id)
+    connection_config = cast(ConnectionConfig, load_item_config(CONNECTION, Path(path)))
+    if (
+        connection_config.cert_requests is None
+        or len(connection_config.cert_requests) == 0
+    ):
+        click.echo("No certificates to process.")
+        return
```
Comment (DavidMinarsch) [inline aea/cli/issue_certificates.py:100]: I don't think we should print this for each connection. Probably make it debug instead.

File: aea/cli/issue_certificates.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            public_key is not None,
+            "Internal error - one of key_identifier or public_key must be not None.",
+        )
+    public_key_bytes = public_key.encode("ascii")
+    identifier = cert_request.identifier.encode("ascii")
+    not_before = cert_request.not_before_string.encode("ascii")
+    not_after = cert_request.not_after_string.encode("ascii")
+    crytpo_private_key_path = ctx.agent_config.connection_private_key_paths.read(
+        ledger_id
+    )
```
Comment (DavidMinarsch) [inline aea/cli/issue_certificates.py:81]: No, for signing you use the non connection key. That way we establish the link

File: aea/cli/issue_certificates.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        connection_crypto = crypto_registry.make(
+            key_identifier, private_key_path=connection_private_key_path
+        )
+        public_key = connection_crypto.public_key
+    else:
+        public_key = cast(str, cert_request.public_key)
+        enforce(
+            public_key is not None,
+            "Internal error - one of key_identifier or public_key must be not None.",
+        )
```
Comment (DavidMinarsch) [inline aea/cli/issue_certificates.py:74]: correct!

File: tests/data/dummy_connection/connection.yaml (hunk 3 above)
Comment (DavidMinarsch) [reply tests/data/dummy_connection/connection.yaml:36]: but we also have `ledger_id`....

**Review by DavidMinarsch — APPROVED** [review]

File: aea/cli/issue_certificates.py (hunk 8 above)
Comment (marcofavorito) [reply aea/cli/issue_certificates.py:81]: yes, that was a typo, sorry

File: tests/data/dummy_connection/connection.yaml (hunk 3 above)
Comment (marcofavorito) [reply tests/data/dummy_connection/connection.yaml:36]: `public_key` is:
- either a public key
- or a key identifier to get the connection key from `connection_key_paths`. 

- Does that make sense?

File: aea/configurations/base.py (hunk 5 above)
Comment (marcofavorito) [reply aea/configurations/base.py]: how?

**Review by DavidMinarsch — APPROVED** [review]
LGTM, thanks!

**Review by lrahmani — APPROVED** [review]
LGTM. 
In subsequent pr we need to address some issues like (1) the need for time in addition to date for `not_before` and `not_after` fields  and (2) adding utilities such as `is_valid` to the produced certificate

**Pull request merged**